src/lib.rs

```diff
@@ -41,52 +41,74 @@ where
     StorageT: AsPrimitive<T> + PrimInt + Unsigned,
 {
     /// Constructs a new `PackedVec` from `vec` (with a user-defined backing storage type).
     pub fn new_with_storaget(vec: Vec<T>) -> PackedVec<T, StorageT> {
         if size_of::<T>() > size_of::<StorageT>() {
             panic!("The backing storage type must be the same size or bigger as the stored integer size.");
         }
-        let m = vec.iter().max();
-        // If the input vector was empty, or if it consisted entirely of zeros, we don't need to
-        // fill the backing storage with anything.
-        if m.is_none() || *m.unwrap() == T::zero() {
+
+        // If the input vector was empty, we don't need to fill the backing storage with anything.
+        if vec.is_empty() {
             return PackedVec {
                 len: vec.len(),
                 bits: vec![],
-                item_width: 0,
-                phantom: PhantomData,
+                bwidth: 0,
+                min: T::zero(), // This value will never be used
             };
         };
 
-        let item_width = i_log2(*m.unwrap());
+        // We now want to find the difference between the biggest and smallest elements so that we
+        // can pack things down as far as possible. In other words, if we have a vec [0, 4] the
+        // range is 4 and we need 2 bits to represent that range; if we have a vec [-1, 1] the
+        // range is 2 and we need 1 bit to represent that range.
+        //
+        // Perhaps surprisingly, it's faster to call iter().max() and iter().min() separately than
+        // it is to write a single for loop over the input vector. This is probably because of the
+        // overhead of an iterator, and perhaps because max() and min() use CPU vector
+        // instructions.
+        let max = *vec.iter().max().unwrap();
+        let min = *vec.iter().min().unwrap();
+        // If the input vector consisted entirely of the same element, we don't need to fill the
+        // backing storage with anything.
+        if max == min {
+            return PackedVec {
+                len: vec.len(),
+                bits: vec![],
+                bwidth: 0,
+                // By definition "min" will be the value
+                min,
+            };
+        };
+        let bwidth = i_log2(delta(min, max));
+
         let mut bit_vec = vec![];
         let mut buf = StorageT::zero();
         let mut bit_count: usize = 0;
-        for item in &vec {
-            let item: StorageT = (*item).as_();
-            if bit_count + item_width < num_bits::<StorageT>() {
-                let shifted_item = item << num_bits::<StorageT>() - (item_width + bit_count);
-                buf = buf | shifted_item;
-                bit_count += item_width;
+        for &e in &vec {
+            let e = delta(min, e);
+            if bit_count + bwidth < num_bits::<StorageT>() {
+                let shifted_e = e << num_bits::<StorageT>() - (bwidth + bit_count);
```

Comment:
I see, so `bits` now stores the distance between each element and the minimum value, which means we can reconstruct the original vector using `bits` and `min`.

Reply:
Correct.